### PR TITLE
Added solution to task on immutable classes

### PR DESCRIPTION
**Pull request by MichaelShevchenko**

### CODE REVIEW

File: src/main/java/core/basesyntax/Car.java [hunk 1]
```diff
@@ -1,13 +1,95 @@
...
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        List<Wheel> newWheels = new ArrayList<>();
+        for (Wheel oneWheel : wheels) {
+            newWheels.add(oneWheel.clone());
+        }
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Car.java:21]: let's move it to the separate private method

File: src/main/java/core/basesyntax/Car.java [hunk 2]
```diff
@@ -1,13 +1,95 @@
...
+        this.wheels = newWheels;
+        this.engine = (engine == null) ? null : new Engine(engine.getHorsePower(),
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Car.java:23]: let's move it to the separate private method

```suggestion
        this.engine = (engine == null) ? null : engine.clone()
```

File: src/main/java/core/basesyntax/Car.java [hunk 3]
```diff
@@ -1,13 +1,95 @@
...
+                                                                engine.getMaker());
+    }
+
+    public int getYear() {
+        return year;
+    }
+
+    public Car changeYear(int year) {
+        return new Car(year, color, new ArrayList<>(wheels),
+                            new Engine(engine.getHorsePower(), engine.getMaker()));
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Car.java:33]: let's call engin.clone() everywhere

File: src/main/java/core/basesyntax/Car.java [hunk 4]
```diff
@@ -1,13 +1,95 @@
...
+        this.engine = (engine == null) ? null : new Engine(engine.getHorsePower(),
+                                                                engine.getMaker());
+    }
+
+    public int getYear() {
+        return year;
+    }
+
+    public Car changeYear(int year) {
+        return new Car(year, color, new ArrayList<>(wheels),
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Car.java:32]:
```suggestion
        return new Car(year, color, wheels,
```

File: src/main/java/core/basesyntax/Car.java [hunk 5]
```diff
@@ -1,13 +1,95 @@
...
+        return color;
+    }
+
+    public Car changeColor(String color) {
+        return new Car(year, color, new ArrayList<>(wheels),
+                        new Engine(engine.getHorsePower(), engine.getMaker()));
+    }
+
+    public List<Wheel> getWheels() {
+        return new ArrayList<>(wheels);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Car.java:46]: do we need to clone all elements at the list?

File: src/main/java/core/basesyntax/Car.java [hunk 6]
```diff
@@ -1,13 +1,95 @@
...
+                                                            engine.getMaker()));
+    }
+
+    public Engine getEngine() {
+        return (engine == null) ? null : new Engine(engine.getHorsePower(), engine.getMaker());
+    }
+
+    public Car changeEngine(Engine engine) {
+        Engine newEngine = engine;
+        return new Car(year, color, new ArrayList<>(wheels), newEngine);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Car.java:62]:
```suggestion
        return new Car(year, color, new ArrayList<>(wheels), engine);
```

Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/Car.java:62]: done

File: src/main/java/core/basesyntax/Car.java (hunk 4 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/Car.java:32]: done

File: src/main/java/core/basesyntax/Car.java (hunk 3 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/Car.java:33]: done

File: src/main/java/core/basesyntax/Car.java (hunk 5 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/Car.java:46]: done

File: src/main/java/core/basesyntax/Car.java (hunk 2 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/Car.java:23]: done

File: src/main/java/core/basesyntax/Car.java (hunk 1 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/Car.java:21]: done

File: src/main/java/core/basesyntax/Car.java [hunk 7]
```diff
@@ -1,13 +1,98 @@
...
+
+    private List<Wheel> wheelsPackage(List<Wheel> wheels) {
+        List<Wheel> newWheels = new ArrayList<>();
+        for (Wheel oneWheel : wheels) {
+            newWheels.add(oneWheel.clone());
+        }
+        return newWheels;
+    }
+
+    private Engine installEngine(Engine engine) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:95]:
```suggestion
    private Engine copyEngine(Engine engine) {
```

File: src/main/java/core/basesyntax/Car.java [hunk 8]
```diff
@@ -1,13 +1,98 @@
...
+    public int hashCode() {
+        int result = year;
+        int prime = 41;
+        result = prime * result + (color != null ? color.hashCode() : 0);
+        result = prime * result + (wheels != null ? wheels.hashCode() : 0);
+        result = prime * result + (engine != null ? engine.hashCode() : 0);
+        return result;
+    }
+
+    private List<Wheel> wheelsPackage(List<Wheel> wheels) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:87]:
```suggestion
    private List<Wheel> copyWheels(List<Wheel> wheels) {
```

File: src/main/java/core/basesyntax/Car.java [hunk 9]
```diff
@@ -1,13 +1,98 @@
...
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(year, color, wheels, engine);
+    }
+
+    @Override
+    public Car clone() {
+        return new Car(year, color, wheels, engine.clone());
+    }
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:59]: this class is immutable :)

File: src/main/java/core/basesyntax/Car.java [hunk 10]
```diff
@@ -1,13 +1,98 @@
...
+        this.wheels = wheelsPackage(wheels);
+        this.engine = installEngine(engine);
+    }
+
+    public int getYear() {
+        return year;
+    }
+
+    public Car changeYear(int year) {
+        return new Car(year, color, wheels, engine.clone());
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:27]:
```suggestion
        return new Car(year, color, wheels, engine);
```
in the constructor you will make a copy of the `engine`

File: src/main/java/core/basesyntax/Car.java [hunk 11]
```diff
@@ -1,13 +1,98 @@
...
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    public Car changeColor(String color) {
+        return new Car(year, color, wheels, engine.clone());
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:35]: same

File: src/main/java/core/basesyntax/Car.java [hunk 12]
```diff
@@ -1,13 +1,98 @@
...
+    }
+
+    public List<Wheel> getWheels() {
+        return wheelsPackage(wheels);
+    }
+
+    public Car addWheel(Wheel wheel) {
+        List<Wheel> newWheels = new ArrayList<>(wheels);
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:43]:
```suggestion
        List<Wheel> newWheels = wheelsPackage(wheels);
```

File: src/main/java/core/basesyntax/Car.java [hunk 13]
```diff
@@ -1,13 +1,98 @@
...
+        newWheels.add(wheel);
+        return new Car(year, color, newWheels, engine.clone());
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:45]:
```suggestion
        return new Car(year, color, newWheels, engine);
```

File: src/main/java/core/basesyntax/Car.java (hunk 10 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/Car.java:27]: done

File: src/main/java/core/basesyntax/Car.java (hunk 11 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/Car.java:35]: done

File: src/main/java/core/basesyntax/Car.java (hunk 13 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/Car.java:45]: done

File: src/main/java/core/basesyntax/Car.java (hunk 8 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/Car.java:87]: done

File: src/main/java/core/basesyntax/Car.java (hunk 7 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/Car.java:95]: done

File: src/main/java/core/basesyntax/Car.java (hunk 9 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/Car.java:59]: took in mind!)

File: src/main/java/core/basesyntax/Car.java (hunk 12 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/Car.java]: done

**Pull request closed without merging**